### PR TITLE
Account sign up

### PR DESCRIPTION
**Pull request by samtechnet**
dist folder deleted

### CODE REVIEW

File: src/controller/auth/verifyAuth.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { NextFunction, Request, Response } from 'express';
+import AppError from "../../services/errorHandlers/errors";
+import { client } from "../../services/database/database";
+
+const verifyToken = async (req: Request, res: Response, next: NextFunction) => { 
+    try {
+        const authentication = req.headers
+        console.log(authentication)
```
Comment (forestbaba) [inline src/controller/auth/verifyAuth.ts:8]: Please remove all console log

File: src/controller/auth/verifyAuth.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import AppError from "../../services/errorHandlers/errors";
+import { client } from "../../services/database/database";
+
+const verifyToken = async (req: Request, res: Response, next: NextFunction) => { 
+    try {
+        const authentication = req.headers
+        console.log(authentication)
+
+        const cookie = req.cookies
+        console.log(cookie)
```
Comment (forestbaba) [inline src/controller/auth/verifyAuth.ts:11]: Here too

File: src/controller/user.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        
+        const str = usercheck.toString()
+        
+        if (str=== "false") {
+            return res.status(422).json({
+                success: false,
+                error: "User account is not active, Kindly activate account"
+           })
+        };
+        const user = await Users.findOne({ email });
```
Comment (forestbaba) [inline src/controller/user.ts:102]: Let this be handled in the service

File: src/controller/user.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            
+        }     
+    } catch (error) {
+        return res.send(new AppError(`something went wrong here is the error ${error}`, 500));
+    } finally {
+        await client.close();
+    };
+    
+}
+const signIn = async (req: Request, res: Response, next: NextFunction) => {
```
Comment (forestbaba) [inline src/controller/user.ts:87]: Can you make this endpoint just to do one things alone which is sign in??...Can all checks be done in the validator and service class or function?

File: src/controller/user.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        if (hashedPassword=== user.password) {
+            let token = generateAccessToken(user)
+            let profile = {
+              email: user.email,
+              role: user.role,
+              name: user.firstName +" "+ user.lastName,
+            };
+            let result = {
+              user: profile,
+              //token: token,
```
Comment (forestbaba) [inline src/controller/user.ts:113]: Remove dead codes

File: src/controller/user.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    };
+};
+
+const activateAccount= async (req: Request, res: Response, next:NextFunction) => {
+        const { code}=req.body
+    try {
+        await client.connect();
+        
+        isCodeactive;
+        const user = await Users.findOne({ verificationCode: Number(code) })
```
Comment (forestbaba) [inline src/controller/user.ts:147]: Database query should be done in repository class or service class

File: src/controller/user.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                message: "Invalid code",
+                success: false,
+            })
+        }else if (user.isEmailVerified) {
+            return res.status(409).json({
+              message: "Email already verified",
+              success: false,
+            });
+        }
+        const modify = await Users.findOneAndUpdate({ verificationCode: user.verificationCode }, { $set: { isEmailVerified: true } });
```
Comment (forestbaba) [inline src/controller/user.ts:160]: Here too

File: src/controller/user.ts [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        return next(new AppError(`something went wrong here is the error ${error}`, 500));
+    } finally {
+        await client.close();
+    };
+}
+
+const geAllUsers=async (req: Request, res: Response, next:NextFunction) => {
+    try {
+        await client.connect();
+        const allUsers = await Users.find({}).toArray()
```
Comment (forestbaba) [inline src/controller/user.ts:178]: See here

File: src/mailers/senders.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        html: output
+      }
+    transporter.sendMail(mailOptions, (error, info) => {
+        if (error) {
+            return console.log(error);
+        }
+        console.log('Message sent: %s', info.messageId);
+        console.log('preview URL: %s', nodemailer.getTestMessageUrl(info))
+    });
+    // const email = new Email({
```
Comment (forestbaba) [inline src/mailers/senders.ts:61]: Remove dead codes

**Review by samtechnet — COMMENTED** [review]
updated code base

File: src/services/database/database.ts [hunk 10]
```diff
@@ -1,28 +1,164 @@
...
+      password,
+      role,
+      isEmailVerified,
+    };
   }
-  export { client,uri };
\ No newline at end of file
+
+  async createOne(user: any) {
+    await client.connect();
```
Comment (forestbaba) [inline src/services/database/database.ts:120]: Why are u connecting always?

File: src/services/database/database.ts [hunk 11]
```diff
@@ -1,28 +1,164 @@
...
+});
+
+if (ENV === "dev") {
+  console.log("I am in dev mode");
+}
+if (ENV === "prod") {
+  console.log("I am in production mode");
+}
+
+class DatabaseQuery {
```
Comment (forestbaba) [inline src/services/database/database.ts:29]: Repository class and config class can stay separately. I'm wandering why everything is in the same class. Makes it look messy

File: src/services/database/database.ts [hunk 12]
```diff
@@ -1,28 +1,164 @@
...
-  export { client,uri };
\ No newline at end of file
+
+  async createOne(user: any) {
+    await client.connect();
+    const data = await DatabaseQuery.Users.insertOne(user);
+    await client.close();
+    return data;
+  }
+  async findByCode(code: string): Promise<string | any> {
```
Comment (forestbaba) [inline src/services/database/database.ts:125]: Try as much as possible to communicate what a function do by their name...For instance u said _findByCode_ find what?
You could have simply use something like:
_findUserByCode_ or _findUser_ then let the argument be an object so i can be dynamic and do alot

File: src/services/database/database.ts [hunk 13]
```diff
@@ -1,28 +1,164 @@
...
+        isEmailVerified,
+      };
+    }
+    if (data === null) {
+      let isNotValid = "isNotValid";
+      return isNotValid;
+    }
+    }
+    async getAll() {
+        await client.connect();
```
Comment (forestbaba) [inline src/services/database/database.ts:156]: But why are u connecting always to make a request?

File: src/services/util.ts [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    let hashKey = crypto
+      .createHmac("SHA256", SecretKey)
+      .update(email)
+      .digest("base64")
+      .toString();
+    return hashKey;
+  }
+  async generateAccessToken(user: any) {
+    const jwsToken = String(process.env.AccessToken);
+    const UserData = new DatabaseQuery();
```
Comment (forestbaba) [inline src/services/util.ts:38]: UserDate as used here is not a good naming convention. Stick to camel case

File: src/middleware/validateRequest.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
+import express, { Request, Response, NextFunction } from "express";
```
Comment (forestbaba) [inline src/middleware/validateRequest.ts:1]: The validator file you added to controller folder should actually be in the same folder as middleware...Why did you keep it there though?

File: src/routes/userRoutes.ts [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+import {createAccountLimiter} from "../middleware/validateRequest"
+
+const user_routes = (app: express.Application) => {
+  app.post("/api/signUp",createAccountLimiter, signUpValidationRules(), validate, signUp);
+  app.post("/api/signIn", validationRules(), validate, signIn);
+  app.post("/api/activateAccount",otpValidate(),validate, activateAccount);
+  app.get("/api/geAllUsers", geAllUsers);
+};
+
+const server = {
```
Comment (forestbaba) [inline src/routes/userRoutes.ts:23]: Wait, What's happening here downwards?

File: src/server.ts [hunk 17]
```diff
@@ -14,7 +19,33 @@ app.use(bodyParser.json());
 
 app.use(cors());
 app.use(express.json());
+app.use(cookieParser());
+app.disable('x-powered-by');
 
+app.use("/documentations", swaggerDoc.serve);
+app.use("/documentations", swaggerDoc.setup(swaggerDocumentation));
+
+const limiter = rateLimit({
```
Comment (forestbaba) [inline src/server.ts:28]: Can i know why you're using limiter?

**Review by samtechnet — COMMENTED** [review]
ready for review

**Review by forestbaba — APPROVED** [review]

**Pull request merged**